src/tracker.rs

```diff
@@ -0,0 +1,300 @@
+                mut seen,
+                mut cardinality_count,
+                actions,
+            } => {
+                let cardinality_grown = Self::update(&mut seen, metric, tags.rewind());
+                if cardinality_grown {
+                    cardinality_count += 1;
+                    if cardinality_count >= self.cardinality_threshold {
+                        drop(lock);
+                        Self::do_actions(dd, seen, actions, metric, tags.rewind());
+                        return;
+                    }
+                }
+                // Rebuild the running state if we didn't run the actions
+                *lock = TrackerState::Running {
+                    seen,
+                    cardinality_count,
+                    actions,
+                };
+            }
+            TrackerState::Done => {}
+        };
+    }
+
+    fn update(seen: &mut BTreeMap<String, Vec<BTreeSet<String>>>, metric: &str, tags: &mut RewindableTagsIter) -> bool {
+        let seen_tag_set = match seen.get_mut(metric) {
+            Some(seen_tag_set) => seen_tag_set,
+            None => {
+                seen.insert(
+                    metric.to_string(),
+                    vec![BTreeSet::from_iter(tags.rewind().map(|tag| tag.to_string()))],
+                );
+                return true;
+            }
+        };
+        let set_is_novel = seen_tag_set
+            .iter()
+            .all(|tag_set| tag_set.len() != tags.len() || tags.rewind().any(|tag| !tag_set.contains(tag)));
+        if set_is_novel {
+            seen_tag_set.push(BTreeSet::from_iter(tags.map(|tag| tag.to_string())));
+        };
+        set_is_novel
+    }
+
+    fn do_actions(
+        dd: &impl DogstatsdClient,
+        seen: BTreeMap<String, Vec<BTreeSet<String>>>,
+        actions: Vec<ThresholdAction>,
+        metric: &str,
+        tags: &mut RewindableTagsIter,
+    ) {
+        let event_tags = seen
+            .iter()
+            .map(|(metric, tags)| format!("{}:{}", metric, tags.len()))
+            .collect::<Vec<_>>();
+
+        let tags = tags.rewind().collect::<Vec<&str>>();
+
+        for action in actions {
+            match action {
+                ThresholdAction::Event { title, text } => dd.event(&title, &text, &event_tags),
+                ThresholdAction::Custom(mut action) => {
+                    action(metric, &tags);
+                }
+            }
+        }
+    }
+}
+
+/// Actions that define what the tracker will do when the custom metric threshold is passed.
+/// A user may define any number of these, and by default none are taken.
+enum ThresholdAction {
+    /// Emit an event. The count of unique tag sets, per metric, is provided as the tags
+    Event { title: String, text: String },
+
+    /// Take some custom action. The function will be passed the metric name and tags
+    Custom(ThresholdCustomAction),
+}
+
+type ThresholdCustomAction = Box<dyn FnMut(&str, &[&str]) + Send + Sync>;
```

Comment:
The user doesn't need to Box `fn`s already, since this type is used only internally by the public method `with_custom_action` which boxes the `fn` passed by the user.